### DISCUSSION TITLE
angular 15 ?

### DISCUSSION THREAD

**Discussion by memory660** [stalniy/casl#724] [General]
on angular 15, Ability is deprecated
will there be a version 15?

**Comment by TheDelta** [comment]
It works fine on angular 15, I think the documentation is just outdated and Ability is not needed anymore.

Just remove any occurrence of `Ability` and replace it with `PureAbility`.

In app.module, instead of `new Ability()` use one of the create methods, for example the default Mongo Adapter. (for reference https://casl.js.org/v6/en/advanced/typescript)

```typescript
import { NgModule } from '@angular/core';
import { AbilityModule } from '@casl/angular';
import { PureAbility, createMongoAbility } from '@casl/ability';

@NgModule({
  imports: [
    // other modules
    AbilityModule
  ],
  providers: [
    { provide: Ability, useValue: createMongoAbility() },
  ]
  // other properties
})
export class AppModule {}
```
(Ability is not needed anymore, the docs seem a bit outdated IMHO)

Also instead of using `Ability` in your services, you must use `PureAbility`.
If you use a custom type or in AppModules use a custom class, factory, etc. then you must inject it:

```typescript
@Injectable({ provideIn: 'root' })
export class Session {
  constructor(
    // use default PureAbility (without typing help)
    // private readonly ability: PureAbility,
    // or if you have a custom ability type, for better typesafity
    @Inject(PureAbility)
    private readonly ability: AppAbility
  ) {}

  private updateAbility(user): void {
    // default
    // const { can, rules } = new AbilityBuilder(createMongoAbility);
    // or type based
    const { can, rules } = new AbilityBuilder<AppAbility>(createMongoAbility);

    if (user.role === 'admin') {
      can('manage', 'all');
    } else {
      can('read', 'all');
    }

    this.ability.update(rules);
  }
}
```

**Full example** https://stackblitz.com/edit/angular-mbdvmu?file=src/main.ts

(NOTE: I'm relatively new to this library)
[Edited: fixed example code]

**Comment by stalniy** [comment]
@TheDelta is right! Thank you for helping to explain this. 

Ability is deprecated because its main purpose can be translated to factory function. In the next major version I plan to rename PureAbility to Ability. This is explained in jsdoc deprecation description which you should see in your IDE

**Comment by memory660** [comment]
thank you for your reply.
here are the errors when I try to adapt your code to mine

![ab1](https://user-images.githubusercontent.com/95366869/224329389-d2f783d4-2417-42af-8d65-d9e60d1ec706.jpg)
![ab2](https://user-images.githubusercontent.com/95366869/224329418-0cba25e2-7b6c-464f-9097-112e21386f62.jpg)

`
  constructor(
    private readonly ability: PureAbility,
  ) {}
`

**Comment by memory660** [comment]
solution:

```ts
import 'zone.js/dist/zone';
import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { bootstrapApplication } from '@angular/platform-browser';
import { AbilityModule } from '@casl/angular';
import { AbilityBuilder, createMongoAbility, PureAbility } from '@casl/ability';

@Component({
  selector: 'my-app',
  standalone: true,
  imports: [CommonModule, AbilityModule],
  providers: [{ provide: PureAbility, useValue: createMongoAbility() }],
  template: `
  <div *ngIf="'read' | able: 'Article'">
    read Article ok
  </div>
  `,
})
export class App {

  constructor(private ability: PureAbility) {
    const { can, rules, build } = new AbilityBuilder(PureAbility);

    const abilityConf = new PureAbility<['read' | 'update', 'Article']>([
      { action: 'read', subject: 'Article' },
      { action: 'update', subject: 'Article' },
    ]);

    this.ability.update(abilityConf.rules);

    console.log('📢 [update Article ?]', ability.can('update', 'Article'));
  }
}

bootstrapApplication(App);
```

**Reply by TheDelta** [reply]
Both of your messages seem like a lack of understanding TypeScript and CASL (no offense).
Not a good solution, you define an `AbilityBuilder`, but you don't even use it?!.

Big sorry that my example code was not solid (going to update it to prevent confusion for other users).
Please refer to this playground how to do a mongo ability (for prisma or others it's almost identical, refer to the documentation).

https://stackblitz.com/edit/angular-mbdvmu?file=src/main.ts

**Reply by memory660** [reply]
great,  thanks !     (you've unlocked me on this topic, I'll be able to use it on the business project!)

```html
Can Write Product: {{ UserActions.WRITE | ablePure: 'product' | async }}
```
I can't find where WRITE and PRODUCT have been associated in the code?

here :
```ts
    if (user.isAdmin) {
      can(UserActions.MANAGE, 'all');
    } else {
      can(UserActions.READ, 'product');
    }

//  there not : can(UserActions.WRITE, 'product'); ???????????
```
